> @tklauser leases `false` or `true` for `UpdateKeySync` from `clustermesh-apiserver/vmmanager.go`?
> 
> ```
> Error: level=warning msg="[runner] Can't run linter goanalysis_metalinter: err113: analysis skipped: errors in package: [/home/runner/work/cilium/cilium/clustermesh-apiserver/vmmanager.go:194:77: too few arguments in call to ciliumNodeRegisterStore.UpdateKeySync]"
> ```

The respective `UpdateSyncKey` call was removed in `master` by #18957 (commit 90c0d62c724469dee07853f8d104eaf167f9c24e) and replaced by a direct call to `kvstore.Client().UpdateIfDifferent` with `lease=true`. So I'd say for the matter of this backport `lease` should be `true` in the call to `UpdateKeySync` (or alternatively backport #18957 to `v1.10` as well).

--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -57,6 +57,7 @@ cilium-agent [flags]
       --cluster-id int                                          Unique identifier of the cluster
       --cluster-name string                                     Name of the cluster (default "default")
       --clustermesh-config string                               Path to the ClusterMesh configuration directory
+      --cni-chaining-mode string                                Enable CNI chaining with the specified plugin
       --config string                                           Configuration file (default "$HOME/ciliumd.yaml")
       --config-dir string                                       Configuration directory that contains a file for each option
       --conntrack-gc-interval duration                          Overwrite the connection-tracking garbage collection interval

--- a/daemon/cmd/daemon_main.go
+++ b/daemon/cmd/daemon_main.go
@@ -476,6 +476,9 @@ func initializeFlags() {
 	flags.String(option.IPAM, ipamOption.IPAMClusterPool, "Backend to use for IPAM")
 	option.BindEnv(option.IPAM)
 
+	flags.String(option.CNIChainingMode, "", "Enable CNI chaining with the specified plugin")
+	option.BindEnv(option.CNIChainingMode)
+
 	flags.String(option.IPv4Range, AutoCIDR, "Per-node IPv4 endpoint prefix, e.g. 10.16.0.0/16")
 	option.BindEnv(option.IPv4Range)
 

--- a/daemon/cmd/ipam.go
+++ b/daemon/cmd/ipam.go
@@ -254,6 +254,15 @@ func (d *Daemon) allocateDatapathIPs(family datapath.NodeAddressingFamily) (rout
 		}
 		routerIP = result.IP
 	}
+
+	// Coalescing multiple CIDRs. GH #18868
+	if option.Config.EnableIPv4Masquerade &&
+		option.Config.IPAM == ipamOption.IPAMENI &&
+		result != nil &&
+		len(result.CIDRs) > 0 {
+		result.CIDRs = coalesceCIDRs(result.CIDRs)
+	}
+
 	if (option.Config.IPAM == ipamOption.IPAMENI ||
 		option.Config.IPAM == ipamOption.IPAMAlibabaCloud ||
 		option.Config.IPAM == ipamOption.IPAMAzure) && result != nil {
@@ -268,10 +277,6 @@ func (d *Daemon) allocateDatapathIPs(family datapath.NodeAddressingFamily) (rout
 		node.SetRouterInfo(routingInfo)
 	}
 
-	// Coalescing multiple CIDRs. GH #18868
-	if result != nil && len(result.CIDRs) > 0 {
-		result.CIDRs = coalesceCIDRs(result.CIDRs)
-	}
 	return
 }
 
@@ -285,7 +290,10 @@ func (d *Daemon) allocateHealthIPs() error {
 			}
 
 			// Coalescing multiple CIDRs. GH #18868
-			if result != nil && len(result.CIDRs) > 0 {
+			if option.Config.EnableIPv4Masquerade &&
+				option.Config.IPAM == ipamOption.IPAMENI &&
+				result != nil &&
+				len(result.CIDRs) > 0 {
 				result.CIDRs = coalesceCIDRs(result.CIDRs)
 			}
 
@@ -313,7 +321,10 @@ func (d *Daemon) allocateHealthIPs() error {
 			}
 
 			// Coalescing multiple CIDRs. GH #18868
-			if result != nil && len(result.CIDRs) > 0 {
+			if option.Config.EnableIPv6Masquerade &&
+				option.Config.IPAM == ipamOption.IPAMENI &&
+				result != nil &&
+				len(result.CIDRs) > 0 {
 				result.CIDRs = coalesceCIDRs(result.CIDRs)
 			}
 

--- a/pkg/datapath/iptables/iptables.go
+++ b/pkg/datapath/iptables/iptables.go
@@ -905,11 +905,21 @@ func (m *IptablesManager) doGetProxyPort(prog iptablesInterface, name string) ui
 }
 
 func getDeliveryInterface(ifName string) string {
-	deliveryInterface := ifName
-	if option.Config.IPAM == ipamOption.IPAMENI || option.Config.IPAM == ipamOption.IPAMAlibabaCloud || option.Config.EnableEndpointRoutes {
-		deliveryInterface = "lxc+"
+	switch {
+	case option.Config.EnableEndpointRoutes:
+		// aws-cni creates container interfaces with names like eni621c0fc8425.
+		if option.Config.CNIChainingMode == "aws-cni" {
+			return "eni+"
+		}
+		return "lxc+"
+
+	case option.Config.IPAM == ipamOption.IPAMENI ||
+		option.Config.IPAM == ipamOption.IPAMAlibabaCloud:
+		return "lxc+"
+
+	default:
+		return ifName
 	}
-	return deliveryInterface
 }
 
 func (m *IptablesManager) installForwardChainRules(ifName, localDeliveryInterface, forwardChain string) error {

--- a/pkg/datapath/loader/base.go
+++ b/pkg/datapath/loader/base.go
@@ -37,7 +37,6 @@ import (
 	datapathOption "github.com/cilium/cilium/pkg/datapath/option"
 	"github.com/cilium/cilium/pkg/datapath/prefilter"
 	"github.com/cilium/cilium/pkg/defaults"
-	iputil "github.com/cilium/cilium/pkg/ip"
 	ipamOption "github.com/cilium/cilium/pkg/ipam/option"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/node"
@@ -167,16 +166,6 @@ func addENIRules(sysSettings []sysctl.Setting, nodeAddressing datapath.NodeAddre
 	routerIP := net.IPNet{
 		IP:   nodeAddressing.IPv4().Router(),
 		Mask: net.CIDRMask(32, 32),
-	}
-
-	cidrs2 := make([]*net.IPNet, 0, 0)
-	for i := range cidrs {
-		cidrs2 = append(cidrs2, &cidrs[i])
-	}
-	resultcidr, _ := iputil.CoalesceCIDRs(cidrs2)
-	cidrs = make([]net.IPNet, 0, len(resultcidr))
-	for _, cidr := range resultcidr {
-		cidrs = append(cidrs, *cidr)
 	}
 
 	for _, cidr := range cidrs {

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -525,6 +525,9 @@ const (
 	// ClusterMeshConfigName is the name of the ClusterMeshConfig option
 	ClusterMeshConfigName = "clustermesh-config"
 
+	// CNIChainingMode configures which CNI plugin Cilium is chained with.
+	CNIChainingMode = "cni-chaining-mode"
+
 	// BPFCompileDebugName is the name of the option to enable BPF compiliation debugging
 	BPFCompileDebugName = "bpf-compile-debug"
 
@@ -1857,6 +1860,9 @@ type DaemonConfig struct {
 	// IPAM is the IPAM method to use
 	IPAM string
 
+	// Enable chaining with another CNI plugin.
+	CNIChainingMode string
+
 	// AutoCreateCiliumNodeResource enables automatic creation of a
 	// CiliumNode resource for the local node
 	AutoCreateCiliumNodeResource bool
@@ -2480,6 +2486,7 @@ func (c *DaemonConfig) Populate() {
 	c.ClusterID = viper.GetInt(ClusterIDName)
 	c.ClusterName = viper.GetString(ClusterName)
 	c.ClusterMeshConfig = viper.GetString(ClusterMeshConfigName)
+	c.CNIChainingMode = viper.GetString(CNIChainingMode)
 	c.DatapathMode = viper.GetString(DatapathMode)
 	c.Debug = viper.GetBool(DebugArg)
 	c.DebugVerbose = viper.GetStringSlice(DebugVerbose)

